core/src/main/java/org/apache/camel/kafkaconnector/CamelSinkTask.java

```diff
@@ -225,15 +230,47 @@ If the CamelMainSupport instance fails to be instantiated (ie.: due to missing c
         }
     }
 
-    private static void mapHeader(Header header, String prefix, Map<String, Object> destination) {
+    private void mapHeader(Header header, String prefix, Map<String, Object> destination) {
         final String key = StringHelper.after(header.key(), prefix, header.key());
         final Schema schema = header.schema();
 
         if (schema.type().equals(Schema.BYTES_SCHEMA.type()) && Objects.equals(schema.name(), Decimal.LOGICAL_NAME)) {
             destination.put(key, Decimal.toLogical(schema, (byte[]) header.value()));
         } else {
-            destination.put(key, header.value());
+            destination.put(key, convertValueFromStruct(header.schema(), header.value()));
+        }
+    }
+
+    private Object convertValueFromStruct(Schema schema, Object value) {
```

Comment:
This also converts message keys, so another name like "convertStructToMap()" makes more sense.